torch_geometric/transforms/rooted_subgraphs.py

```diff
@@ -0,0 +1,213 @@
+            # batched_edge_attr[subgraphs_edges_mapper] shoud be batched_combined_subgraphs_edge_attr
+            return num_edges
+        else:
+            return super().__inc__(key, value, *args, **kwargs)
+
+    def __cat_dim__(self, key, value, *args, **kwargs):
+        if bool(re.search('(combined_subgraphs)', key)):
+            return -1
+        else:
+            return super().__cat_dim__(key, value, *args, **kwargs)
+
+
+def extract_subgraphs(edge_index, num_nodes, num_hops, walk_length=0, p=1, q=1,
+                      repeat=1):
+    if walk_length > 0:
+        node_mask, hop_indicator = random_walk_subgraph(
+            edge_index, num_nodes, walk_length, p=p, q=q, repeat=repeat,
+            cal_hops=True)
+    else:
+        node_mask, hop_indicator = k_hop_subgraph(edge_index, num_nodes,
+                                                  num_hops)
+    edge_mask = node_mask[:, edge_index[0]] & node_mask[:, edge_index[
+        1]]  # N x E dense mask matrix
+    return node_mask, edge_mask, hop_indicator
+
+
+def to_sparse(node_mask, edge_mask, hop_indicator):
+    subgraphs_nodes = node_mask.nonzero().T
+    subgraphs_edges = edge_mask.nonzero().T
+    if hop_indicator is not None:
+        hop_indicator = hop_indicator[subgraphs_nodes[0], subgraphs_nodes[1]]
+    return subgraphs_nodes, subgraphs_edges,
+
+
+def combine_subgraphs(edge_index, subgraphs_nodes, subgraphs_edges,
+                      num_selected=None, num_nodes=None):
+    if num_selected is None:
+        num_selected = subgraphs_nodes[0][-1] + 1
+    if num_nodes is None:
+        num_nodes = subgraphs_nodes[1].max() + 1
+
+    combined_subgraphs = edge_index[:, subgraphs_edges[1]]
+    node_label_mapper = edge_index.new_full((num_selected, num_nodes), -1)
+    node_label_mapper[subgraphs_nodes[0], subgraphs_nodes[1]] = torch.arange(
+        len(subgraphs_nodes[1]))
+    node_label_mapper = node_label_mapper.reshape(-1)
+
+    inc = torch.arange(num_selected) * num_nodes
+    combined_subgraphs += inc[subgraphs_edges[0]]
+    combined_subgraphs = node_label_mapper[combined_subgraphs]
+    return combined_subgraphs
+
+
+##########################
+# Helpers
+##########################
+from torch_sparse import SparseTensor
+
+
+def k_hop_subgraph(edge_index, num_nodes, num_hops):
+    # return k-hop subgraphs for all nodes in the graph
+    row, col = edge_index
+    sparse_adj = SparseTensor(row=row, col=col,
+                              sparse_sizes=(num_nodes, num_nodes))
+    hop_masks = [
+        torch.eye(num_nodes, dtype=torch.bool, device=edge_index.device)
+    ]  # each one contains <= i hop masks
+    hop_indicator = row.new_full((num_nodes, num_nodes), -1)
+    hop_indicator[hop_masks[0]] = 0
+    for i in range(num_hops):
+        next_mask = sparse_adj.matmul(hop_masks[i].float()) > 0
+        hop_masks.append(next_mask)
+        hop_indicator[(hop_indicator == -1) & next_mask] = i + 1
+    hop_indicator = hop_indicator.T  # N x N
```

Comment:
Why do we need this?

Reply:
this part is to compute the additional feature: distance to centroid. We can remove this part first and add back later.